org/mozilla/jss/tests/TestSSLEngine.java

```diff
@@ -0,0 +1,354 @@
+        assert(server_eng.getHandshakeStatus() == SSLEngineResult.HandshakeStatus.NOT_HANDSHAKING || server_eng.getHandshakeStatus() == SSLEngineResult.HandshakeStatus.FINISHED);
+        assert(client_eng.getStatus().on == 1);
+        assert(server_eng.getStatus().on == 1);
+        assert(client_eng.isInboundDone() == false);
+        assert(server_eng.isInboundDone() == false);
+
+        client_eng.closeOutbound();
+        assert(client_eng.isOutboundDone() == true);
+        sendData(client_eng, server_eng);
+        assert(server_eng.isInboundDone() == true);
+
+        server_eng.closeOutbound();
+        assert(server_eng.isOutboundDone() == true);
+        sendData(server_eng, client_eng);
+        assert(client_eng.isInboundDone() == true);
+    }
+
+    public static void testBasicHandshake(SSLContext ctx, String client_alias, String server_alias) throws Exception {
+        JSSEngine client_eng = (JSSEngine) ctx.createSSLEngine();
+        client_eng.setSSLParameters(createParameters(client_alias));
+        client_eng.setUseClientMode(true);
+        client_eng.beginHandshake();
+
+        System.err.println("client_eng protocols: ");
+        for (String version : client_eng.getEnabledProtocols()) {
+            System.err.println(" - " + version);
+        }
+
+        JSSEngine server_eng = (JSSEngine) ctx.createSSLEngine();
+        server_eng.setSSLParameters(createParameters(server_alias));
+        server_eng.setUseClientMode(false);
+        server_eng.beginHandshake();
+
+        testHandshake(client_eng, server_eng);
+        testPostHandshakeTransfer(client_eng, server_eng);
+        testClose(client_eng, server_eng);
+    }
+
+    public static void testBasicClientServer(String[] args) throws Exception {
+        SSLContext ctx = SSLContext.getDefault();
+        ctx.init(getKMs(), getTMs(), null);
+
+        String client_alias = args[2];
+        String server_alias = args[3];
+
+        testBasicHandshake(ctx, client_alias, server_alias);
+    }
+
+    public static void testNativeClientServer(String[] args) throws Exception {
+        SSLContext ctx = SSLContext.getDefault();
+        ctx.init(getKMs(), new TrustManager[] { new JSSNativeTrustManager() }, null);
+
+        String client_alias = args[2];
+        String server_alias = args[3];
+
+        testBasicHandshake(ctx, client_alias, server_alias);
+    }
+
+    public static void main(String[] args) throws Exception {
+        // Args:
+        //  - nssdb
+        //  - nssdb password
+        //  - client cert
+        //  - server cert
+
+        System.out.println("Initializing CryptoManager...");
+        initialize(args);
+
+        System.out.println("Testing provided instance...");
+        testProvided();
+
+        System.out.println("Testing basic handshake with TMs from provider...");
+        testBasicClientServer(args);
+
+        System.out.println("Testing basic handshake with native TM...");
+        testBasicClientServer(args);
```

Comment:
Should this be testNativeClientServer()?

Reply:
Yeah, I've updated that.